ADR/0027-availability-probe-framework.md

```diff
@@ -0,0 +1,132 @@
+
+### Details
+
+The availability Prometheus metric will be computed for each component based on the exit
+status of the latest execution of the CronJobs evaluating the component's availability.
+Component owners will provide the implementation for each component's CronJobs. By
+adhering to a specific standard
+([see naming convention below](#Probes-Naming-Convention)),
+results will be aggregated into a standardized Prometheus metric to report on
+availability (i.e. component owners will not be required to provide the translation
+mechanism).
+
+It is up for each team to define what it means for its component(s) to be available.
+Fundamentally, a component should be reported as available as long as it's capable of
+providing the service it aims to provide, and unavailable otherwise.
+
+Each team will define CronJobs that will test the availability of their components.
+The Job started by each CronJob will terminate successfully if the test completes
+successfully, and will terminate with an error in case the test fails.
+
+Kube-state-metrics is a Prometheus exporter generating metrics based on the Kubernetes
+API server. It generates
+[Jobs](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/job-metrics.md)
+and
+[CronJob](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/cronjob-metrics.md)
+metrics that will be processed using Prometheus recording rules in order to
+generate AppStudio's availability Prometheus metric.
+
+A single set of rules will be defined globally which will apply for all CronJobs.
+
+The resulting Prometheus metric will identify each probe based on its CronJob name and
+the namespace in which the CronJob was defined.
+
+#### Probes Requirements
+
+A Job running for a probe is required to:
+
+* Evaluate the availability of the component being probed.
+* Exit with status `Failed` if the component was evaluated to be unavailable or with
+  status `Complete` if it was evaluated to be available (referring to the status field
+  of the Kubernetes Job resource).
+* Be able to exhaust its `backoffLimit` by the time it's due to run again (e.g. if the
+  cronjob is defined to run every 10 minutes and can take up to 2 minutes to execute
+  it cannot have a `backoffLimit` larger than 4).
+* Clean up all resources generated during its run.
+
+#### Probes Naming Convention
+
+To allow generating the Prometheus metric only for relevant CronJobs, probe CronJob
+names should have a standardized format:
+
+* `appstudio-probe-<probe_name>`
+
+`appstudio-probe-` being a literal to be used in order to capture only the relevant
+CronJobs, and `<probe_name>` is a variable to be translated to a label in the resulting
+Prometheus metric that will correlate the value to the individual probe or check.
+
+To allow aggregating the Prometheus availability metric per component, the namespaces
+in which the CronJobs will be created should have a standardized format:
+
+* `appstudio-probe-<component_name>`
```

Comment:
We plan on providing an example probe that will require setting up some resources (service account, role binding, etc.) so it will serve as an example for the teams to follow so they don't need to come up with such solution on their own.